d3a_api_client/redis_client_base.py

```diff
@@ -164,31 +165,46 @@ def _publish_and_wait(self, command_type, data):
         return self._wait_and_consume_command_response(command_type, data["transaction_id"])
 
     @registered_connection
-    def offer_energy(self, energy, price):
+    def offer_energy(self, energy: float, price: float, replace_existing: bool = True):
         logging.debug(f"Client tries to place an offer for {energy} kWh at {price} cents.")
-        response = self._publish_and_wait(Commands.OFFER, {"energy": energy, "price": price})
+
+        response = self._publish_and_wait(
+            Commands.OFFER,
+            {'energy': energy, 'price': price, 'replace_existing': replace_existing})
         log_bid_offer_confirmation(response)
+
         return response
 
     @registered_connection
-    def offer_energy_rate(self, energy, rate):
+    def offer_energy_rate(self, energy: float, rate: float, replace_existing: bool = True):
         logging.debug(f"Client tries to place an offer for {energy} kWh at {rate} cents/kWh.")
-        response = self._publish_and_wait(Commands.OFFER, {"energy": energy, "price": rate * energy})
+
+        response = self._publish_and_wait(
+            Commands.OFFER,
+            {'energy': energy, 'price': rate * energy, 'replace_existing': replace_existing})
         log_bid_offer_confirmation(response)
+
         return response
 
-    @registered_connection
-    def bid_energy(self, energy, price):
+    def bid_energy(self, energy: float, price: float, replace_existing: bool = True):
         logging.debug(f"{self.area_id}Client tries to place a bid for {energy} kWh at {price} cents.")
-        response = self._publish_and_wait(Commands.BID, {"energy": energy, "price": price})
+
+        response = self._publish_and_wait(
+            Commands.BID,
+            {'energy': energy, 'price': price, 'replace_existing': replace_existing})
         log_bid_offer_confirmation(response)
+
```

Comment:
Also here, the additional two blank lines. IMO they dont improve the readability, only the length of the file.

Reply:
I tend to break code into logically-related parts (see: [PEP 8 - Blank Lines](https://www.python.org/dev/peps/pep-0008/#blank-lines))
> Use blank lines in functions, sparingly, to indicate logical sections.

Of course, the _sparingly_ can be interpreted :D

In this case, I've "seen" three parts:
- initial part for logging (before anything is done)
- the actual logic of the method and its logging
- the `return` statement (I generally add a blank line before return statements, unless they're one-liners).
 
This helps me to clearly identify the different parts of a method and to visually understand when a chunk of code has "done its part". I understand your concerns about the length of the file, but I prefer the risk of being unnecessarily verbose to the risk of being overly cryptic (not in this case, but as a general rule). It's very subjective though.

On a separate note: if the file has too many lines, most of the time it's because of a design issue (for example, two classes are needed instead of one).

Reply:
you are of course right, especially with the latter comment.
If we would have time i would have proposed to add a refactoring story for this for each repo we have, but we dont, so yeah, will keep your explanations in mind from now on. Thanks a lot for bearing with me!

Reply:
Of course, there's a time/quality balance to keep in mind, and code quality can/should sometimes be sacrificed depending on the time availability (e.g. deadlines or low resources) :)